esti/s3_gateway_test.go

```diff
@@ -144,90 +144,173 @@ func TestS3ReadObject(t *testing.T) {
-			if err == nil {
-				t.Errorf("Successfully read \"%s\" from nonexistent path %s", got, badPath)
-			}
-			s3ErrorResponse := minio.ToErrorResponse(err)
-			if s3ErrorResponse.StatusCode != 404 {
-				t.Errorf("Got %+v [%d] on reading when expecting Not Found [404]",
-					s3ErrorResponse, s3ErrorResponse.StatusCode)
-			}
-		})
+		resp, err := http.Get(preSignedURL.String())
+		if err != nil {
+			t.Fatalf("http.Get %s: %s", preSignedURL.String(), err)
+		}
+		defer func() { _ = resp.Body.Close() }()
+		contentLength := resp.Header.Get("Content-Length")
+		if contentLength == "" {
+			t.Errorf("Expected Content-Length header to be set")
+		}
 	})
 
-	t.Run("HeadObject", func(t *testing.T) {
-		t.Run("Exists", func(t *testing.T) {
-			info, err := client.StatObject(ctx, repo, goodPath, minio.StatObjectOptions{})
-			if err != nil {
-				t.Errorf("client.StatObject(%s, %s): %s", repo, goodPath, err)
-			}
-			verifyObjectInfo(t, info, len(contents))
+	t.Run("get_no_physical_object", func(t *testing.T) {
+		blockStoreType := viper.GetString("blockstore_type")
+		if blockStoreType != "s3" {
+			t.Skip("Skipping test - blockstore type is not s3")
+		}
+		awsAccessKeyID := viper.GetString("aws_access_key_id")
+		awsSecretAccessKey := viper.GetString("aws_secret_access_key")
+		if awsAccessKeyID == "" || awsSecretAccessKey == "" {
+			t.Skip("Skipping test - AWS credentials not set")
+		}
+
+		// prepare entry with no physical object under it
+		const (
+			objPath   = "no-physical-object"
+			objS3Path = "main/" + objPath
+		)
+		_, err = minioClient.PutObject(ctx, repo, objS3Path, strings.NewReader(""), int64(0), minio.PutObjectOptions{})
+		if err != nil {
+			t.Errorf("PutObject(%s, %s): %s", repo, objS3Path, err)
+		}
+		statResp, err := client.StatObjectWithResponse(ctx, repo, "main", &apigen.StatObjectParams{
+			Path:    objPath,
+			Presign: swag.Bool(false),
 		})
-		t.Run("Doesn't exist", func(t *testing.T) {
-			info, err := client.StatObject(ctx, repo, badPath, minio.StatObjectOptions{})
-			if err == nil {
-				t.Errorf("client.StatObject(%s, %s): expected an error but got %+v", repo, badPath, info)
-			}
+		if err != nil {
+			t.Fatalf("StatObject(%s, %s): %s", repo, objPath, err)
+		}
+		if statResp.JSON200 == nil {
+			t.Fatalf("StatObject(%s, %s): got nil response, status: %s", repo, objPath, statResp.Status())
+		}
+		physicalAddress := statResp.JSON200.PhysicalAddress
 
-			s3ErrorResponse := minio.ToErrorResponse(err)
-			if s3ErrorResponse.StatusCode != 404 {
-				t.Errorf("Got %+v [%d] on reading when expecting Not Found [404]",
-					s3ErrorResponse, s3ErrorResponse.StatusCode)
-			}
+		// setup s3 client and delete the physical object
+		cfg, err := config.LoadDefaultConfig(ctx, config.WithCredentialsProvider(
+			awscredentials.NewStaticCredentialsProvider(awsAccessKeyID, awsSecretAccessKey, "")),
+		)
+		if err != nil {
+			t.Fatalf("failed to load aws configuration, %v", err)
+		}
+		s3Client := s3.NewFromConfig(cfg)
```

Comment:
IIUC this uses minioClient to access lakeFS and s3Client to access the underlying blockstore.  Can we name them for their use rather than their implementation?  So maybe `s3LakeFSClient` and `s3BlockstoreClient`.  Also see no need to use the AWS S3 client here.

Reply:
Renamed to s3UnderlyingClient - this is the only client that access the underlying storage and the name is used in this line and at the call to delete. So I assume there should not be any mix between which client does what.

Reply:
Thanks!  Consider still renaming "minioClient".